sdk/dapp-kit/src/components/SuiClientProvider.tsx

```diff
@@ -2,29 +2,72 @@
 // SPDX-License-Identifier: Apache-2.0
 
 import { SuiClient, getFullnodeUrl } from '@mysten/sui.js/client';
-import { SuiClientContext } from '../hooks/useSuiClient.js';
-import { useMemo } from 'react';
+import type { SuiClientOptions } from '@mysten/sui.js/client';
+import { createContext, useMemo, useState } from 'react';
 
-export interface SuiClientProviderProps {
+type NetworkConfig = SuiClient | SuiClientOptions;
+type NetworkConfigs<T extends NetworkConfig = NetworkConfig> = Record<string, T>;
+
+export interface SuiClientProviderContext {
+	client: SuiClient;
+	queryKey: (key: unknown[]) => unknown[];
+	networks: NetworkConfigs;
+	selectNetwork: (network: string) => void;
+}
+
+export const SuiClientContext = createContext<SuiClientProviderContext | undefined>(undefined);
+
+export interface SuiClientProviderProps<T extends NetworkConfigs> {
+	networks?: T;
+	createClient?: (name: keyof T, config: T[keyof T]) => SuiClient;
+	defaultNetwork?: keyof T;
 	children: React.ReactNode;
-	client?: SuiClient;
-	url?: string;
-	queryKeyPrefix: string;
 }
 
-export const SuiClientProvider = (props: SuiClientProviderProps) => {
-	const ctx = useMemo(() => {
-		const client =
-			props.client ??
-			new SuiClient({
-				url: props.url ?? getFullnodeUrl('devnet'),
-			});
+export function SuiClientProvider<T extends NetworkConfigs>(props: SuiClientProviderProps<T>) {
+	const networks = useMemo(
+		() =>
+			props.networks ??
+			({
+				devnet: { url: getFullnodeUrl('devnet') },
+			} as unknown as T),
+		[props.networks],
+	);
+	const [selectedNetwork, setSelectedNetwork] = useState<keyof T>(
+		props.defaultNetwork ?? Object.keys(networks)[0],
+	);
+
+	const createClient = useMemo(() => {
+		if (props.createClient) {
+			return props.createClient;
+		}
 
+		return (_name: keyof T, config: T[keyof T]) => {
+			if (config instanceof SuiClient) {
+				return config;
+			}
+
+			return new SuiClient(config);
+		};
+	}, [props.createClient]);
+
+	const [client, setClient] = useState<SuiClient>(() => {
+		return createClient(selectedNetwork, networks[selectedNetwork]);
+	});
+
+	const ctx = useMemo((): SuiClientProviderContext => {
```

Comment:
This memo is def correct btw, but I will flag that it's only going to really work if folks memo the inbound props (it's honestly fine if they don't since this should be at a super high-level in the tree and not rerender often anyway though). There's no real way around that, we just can make sure we encourage best behavior through things like good docs and examples.